### PR TITLE
slog: faster and easier on the memory

### PR DESCRIPTION
**Pull request by holiman**
These changes improves the performance of the non-coloured terminal formatting, _quite a lot_. 

```
ok      github.com/ethereum/go-ethereum/log     17.305s
[user@work go-ethereum]$ benchstat slog.before slog.after 
name               old time/op    new time/op    delta
GloggerTerminal-8    9.46µs ± 8%    6.07µs ±20%  -35.85%  (p=0.000 n=10+10)

name               old alloc/op   new alloc/op   delta
GloggerTerminal-8    1.96kB ± 0%    0.23kB ± 0%  -88.47%  (p=0.000 n=10+10)

name               old allocs/op  new allocs/op  delta
GloggerTerminal-8      33.0 ± 0%       5.0 ± 0%  -84.85%  (p=0.000 n=10+10)
```

I tried to _somewhat_ organize the commits, but the it might still be a bit chaotic. Some core insights: 

- The function `terminalHandler.Handl` uses a mutex, and writes all output immediately to 'upstream'. Thus, it can reuse a scratch-buffer every time. 
- This buffer can be propagated internally, making all the internal formatters either write directly to it,
- OR, make  use of the `tmp := buf.AvailableBuffer()` in some cases, where a byte buffer "extra capacity" can be temporarily used. 
- The `slog` package  uses `Attr` by value. It makes sense to minimize operating on them, since iterating / collecting into a new slice, iterating again etc causes copy-on-heap. Better to operate on them only once. 
- If we want to do padding, it's better to copy from a constant `space`-buffer than to invoke `bytes.Repeat` every single time. 

I haven't merged it directly into the slog PR, maybe we should get that fixed and merged first? We can do either way, putting it up here for some feeback.

### CODE REVIEW

File: log/handler.go [hunk 1]
```diff
@@ -277,13 +283,13 @@ func builtinReplace(_ []string, attr slog.Attr) slog.Attr {
...
 		} else {
-			attr.Value = slog.AnyValue(formatLogfmtBigInt(v))
+			attr.Value = slog.AnyValue(string(appendBigInt(nil, v)))
 		}
 	case *uint256.Int:
 		if v == nil {
 			attr.Value = slog.AnyValue("<nil>")
 		} else {
-			attr.Value = slog.AnyValue(formatLogfmtUint256(v))
+			attr.Value = slog.AnyValue(appendU256(nil, v))
```
Comment (namiloh) [inline log/handler.go]: String it

File: log/format.go [hunk 2]
```diff
@@ -116,168 +125,129 @@ func LogfmtFormat() Format {
...
+	// We can use the 'tmp' as a scratch-buffer, to first format the
+	// value, and in a second step do escaping.
+	internal := fmt.Appendf(tmp, "%+v", value)
+	return appendEscapeString(tmp, string(internal))
 }
 
-// FormatLogfmtInt64 formats n with thousand separators.
-func FormatLogfmtInt64(n int64) string {
+// appendFormatLogfmtInt64 formats n with thousand separators into buffer dst.
+func appendInt64(dst []byte, n int64) []byte {
```
Comment (namiloh) [inline log/format.go:212]: Docstring

**Comment by holiman** [comment]
Rebased on latest, also improved some. 
```
$ benchstat slog.before slog.after4 
name               old time/op    new time/op    delta
GloggerTerminal-8    9.10µs ± 9%    4.86µs ±14%  -46.55%  (p=0.000 n=10+10)

name               old alloc/op   new alloc/op   delta
GloggerTerminal-8    1.89kB ± 0%    0.18kB ± 0%  -90.57%  (p=0.000 n=10+10)

name               old allocs/op  new allocs/op  delta
GloggerTerminal-8      30.0 ± 0%       4.0 ± 0%  -86.67%  (p=0.000 n=10+10)
```
(still, let's merge the original one first)

File: log/format.go [hunk 3]
```diff
@@ -44,188 +46,184 @@ type TerminalStringer interface {
...
+			buf.WriteString("\x1b[0m=")
+		} else {
+			buf.Write(appendEscapeString(buf.AvailableBuffer(), attr.Key))
+			buf.WriteByte('=')
 		}
-
-		key := escapeString(attr.Key)
-		rawVal := attr.Value.Any()
-		val := FormatLogfmtValue(rawVal, true)
+		tmp := buf.AvailableBuffer()
```
Comment (jwasinger) [inline log/format.go:109]: `bytes.AvailableBuffer` was introduced in go1.21 .  for now, can we make use of a separate temporary buffer, clearing it on each call to `appendEscapeString`?

Comment (holiman) [reply log/format.go:109]: Good catch! Something temporary yes, lemme think about it...

**Pull request closed without merging**